presto-main/src/main/java/com/facebook/presto/dynamicCatalog/DynamicAddCatalogController.java

```diff
@@ -0,0 +1,150 @@
+import com.facebook.presto.metadata.InternalNodeManager;
+import com.facebook.presto.metadata.StaticCatalogStoreConfig;
+import com.google.common.base.Joiner;
+import com.google.common.base.Splitter;
+import io.airlift.discovery.client.Announcer;
+import io.airlift.discovery.client.ServiceAnnouncement;
+import io.airlift.log.Logger;
+
+import javax.inject.Inject;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import java.io.File;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+import static com.google.common.base.Strings.nullToEmpty;
+import static io.airlift.discovery.client.ServiceAnnouncement.serviceAnnouncement;
+
+@Path("/v1/catalog")
+public class DynamicAddCatalogController
+{
+    private final CatalogManager catalogManager;
+    private final ConnectorManager connectorManager;
+    private final Announcer announcer;
+    private final InternalNodeManager internalNodeManager;
+    private final File catalogConfigurationDir;
+    private static final Logger log = Logger.get(DynamicAddCatalogController.class);
+    private final ResponseParser responseParser;
+
+    private DynamicAddCatalogController(CatalogManager catalogManager, ConnectorManager connectorManager, Announcer announcer, InternalNodeManager internalNodeManager, File catalogConfigurationDir, ResponseParser responseParser)
+    {
+        this.catalogManager = catalogManager;
+        this.connectorManager = connectorManager;
+        this.announcer = announcer;
+        this.internalNodeManager = internalNodeManager;
+        this.catalogConfigurationDir = catalogConfigurationDir;
+        this.responseParser = responseParser;
+    }
+
+    @Inject
+    public DynamicAddCatalogController(CatalogManager catalogManager, ConnectorManager connectorManager, Announcer announcer, InternalNodeManager internalNodeManager, StaticCatalogStoreConfig config, ResponseParser responseParser)
+    {
+        this(catalogManager, connectorManager, announcer, internalNodeManager, config.getCatalogConfigurationDir(), responseParser);
+    }
+
+    @Path("add")
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response addCatalog(CatalogVo catalogVo)
+    {
+        try {
+            log.info("addCatalog : input values are " + catalogVo);
+            ConnectorId connectorId = connectorManager.createConnection(catalogVo.getCatalogName(), catalogVo.getConnectorName(), catalogVo.getProperties());
+            updateConnectorIdAnnouncement(announcer, connectorId, internalNodeManager);
+            log.info("addCatalog() : catalogConfigurationDir: " + catalogConfigurationDir.getAbsolutePath());
+            writeToFile(catalogVo);
```

Comment:
I‘m sorry,  I find it will cause a proplem that 'no query nodes availabled'  because you just write catalog file to coordinator node.

Reply:
There is already a pull request for same here https://github.com/prestodb/presto/pull/12605
You can go through the comments in that PR.
Just to answer your question, we have to call this endpoint from all worker nodes to register the catalog, We can get the connected nodes details while executing the query : ``select * from system.runtime.nodes`` and then register the same catalog to all nodes.  Let me know if that answer your question.